harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,396 @@
+                    }
+                }
+            )
+        return metrics
+
+
+class _Metric:
+    def __init__(
+        self,
+        group: str,
+        metrics: Dict[str, Any],
+        timestamp: datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _ShutdownMessage:
+    pass
+
+
+class _Collector(threading.Thread):
+    def __init__(
+        self,
+        metrics_queue: queue.Queue,
+        collection_interval: int = 1,
+        # xxx: figure out right time for this
+        max_collection_s: int = 300,
+    ):
+        self._collection_interval = collection_interval
+        self._should_exit = False
+        self._last_flush_ts = None
+        self._metrics_queue = metrics_queue
+        self._max_collection_s = max_collection_s
+
+        self._metric_collectors = [
+            _GPU(),
+            _CPU(),
+            _Memory(),
+            _Disk(),
+            _Network(),
+        ]
+        self._should_exit = threading.Event()
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        collection_stop_ts = time.time() + self._max_collection_s
+        while not self._should_exit.is_set():
+            if time.time() > collection_stop_ts:
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(_ShutdownMessage())
+                return
+
+            self._collect_metrics()
+            self._should_exit.wait(self._collection_interval)
+
+    def stop(self) -> None:
+        self._should_exit.set()
+        self._metrics_queue.put(_ShutdownMessage())
+
+    def _collect_metrics(self) -> None:
+        for collector in self._metric_collectors:
+            timestamp = datetime.now(timezone.utc)
+            metrics = collector.collect()
+            if not metrics:
+                continue
+            group_metrics = _Metric(group=collector.group, metrics=metrics, timestamp=timestamp)
+            self._metrics_queue.put(group_metrics)
+
+
+class _Shipper(threading.Thread):
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        metrics_queue: queue.Queue,
+        flush_interval: int = 10,
+    ):
+        self._metrics_queue = metrics_queue
+        self._should_exit = False
+        self._flush_interval = flush_interval
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._metrics: List[_Metric] = []
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        while True:
+            next_flush_ts = time.time() + self._flush_interval
+            # Pop messages from queue until the next scheduled flush time.
+            while time.time() < next_flush_ts:
+                deadline = next_flush_ts - time.time()
```

Comment:
whoa, you got a really sneaky bug here.

Your concept of "now" is changing because each call to `time.time()` is going to change.  So you can end up with `deadline` as a negative value, and `queue.get()` will block forever.

Reply:
I like to set a variable called `now` and reuse it during my time calculation.  Example [here](https://github.com/determined-ai/determined/blob/2dfc4f208dfb1a3afa8ee3c4de14a71216934493/master/static/srv/ship_logs.py#L311-L321).